pandera/schemas.py

```diff
@@ -689,6 +690,181 @@ def to_yaml(self, fp: Union[str, Path] = None):
+                                  name = col,
+                                  checks = new_schema.columns[col].checks,
+                                  nullable = new_schema.columns[col].nullable,
+                                  allow_duplicates = new_schema.columns[col].allow_duplicates,
+                                  coerce = new_schema.columns[col].coerce))
+
+
+        if len(ind_list) == 1:
+            new_schema.index = ind_list[0]
+        elif len(ind_list) > 1:
+            new_schema.index = MultiIndex(ind_list)
+
+        # if drop is True as defaulted, drop the columns moved into the index
+        if drop:
+            new_schema.columns = new_schema.remove_columns(keys).columns
+
+        if not inplace:
+            return new_schema
+        else:
+            self.columns = new_schema.columns
+            self.index = new_schema.index
+
+    def reset_index(self, level: List[str] = None, drop: bool = False, inplace: bool = False):
+        """
+        A method for reseting the :class:`Index` of a :class:`DataFrameSchema`.
+
+        :param level: list of labels
+        :param drop: bool, default True
+        :param append: bool, default False
+        :param inplace: bool, default False
+        :return: a new :class:`DataFrameSchema` with specified column(s) in the index.        
+        
+        """
+        # first check if should be done to self or make copy
+        if inplace:
+            new_schema = self
+        else:
+            new_schema = copy.deepcopy(self)
+        try:
+            assert new_schema.index is not None
+        except AssertionError:
+            raise Exception('There is currently no index set for this schema.')
+
+        # ensure all specified keys are present in the index
+        try:
+            if isinstance(new_schema.index, MultiIndex) and (level is not None):
+                not_in_cols: List[str] = [x for x in level if x not in list(new_schema.index.columns.keys())]
+            elif isinstance(new_schema.index, Index) and (level is not None):
+                not_in_cols: List[str] = [] if ([new_schema.index.name] == level) else level
+            else:
+                not_in_cols:list = []
+            assert not_in_cols == []
+        except AssertionError:
+            raise Exception(f"Keys {not_in_cols} not found in schema columns!")
+
+        # make a list if not None
+        if (level is not None) and (isinstance(level, list) is False):
+            level = list(level)
+        
+        # ensure no duplicates
+        if level is not None:
+            try:
+                dup_cols:List[str] = [x for x in set(level) if level.count(x) > 1]
+                assert dup_cols == []
+            except AssertionError:
+                raise Exception(f"Keys {dup_cols} are duplicated!")
+
+        #
+        additional_columns: list = []
+        new_index = new_schema.index
+        if level is None:
```

Comment:
the conditional branching from lines 817-861 makes the logic a little bit difficult to follow. Consider an alternative approach, something a little flatter like:

```python
        # make sure level is a consistent type, similar to lines 803-804, but
        # now level is guaranteed to be a list
        level = (
            [] if level is None
            else [level] if not isinstance(level, list)
            else level
        )

        # get names of all indexes
        index_levels = (
            [index.name for index in new_schema.index.indexes]
            if isinstance(new_schema.index, MultiIndex)
            else [new_schema.index.name]
        )

        # determine additional columns
        additional_columns = {} if drop else {
            key: new_schema.index.columns[key]
            for key in index_levels if key in level
        }

        # determine new index
        new_index: Optional[Union[Index, MultiIndex]]
        if isinstance(new_schema.index, MultiIndex):
            new_index = MultiIndex([
                index for index in new_schema.index.indexes
                if index.name not in level
            ])
        elif new_schema.index.name not in level:
            new_index = new_schema.index
        else:
            new_index = None

        new_schema = new_schema.add_columns({
            col: new_schema.index.columns[col] for col in additional_columns
        })
        new_schema.index = new_index
```